File: Java/Chapter2/src/Practice/MergeSort.java
package Practice;

/**
 * Created by Mingyuan Li on 2016/2/10.
 * Package name: Practice.
 * Project name: Chapter2.
 */
public class MergeSort
{
    private int[] array;
    private int[] tempArray;

    public static void main(String a[]){

        int[] inputArr = {45,23,11,89,77,98,4,28,65,43};
        MergeSort mms = new MergeSort();
        mms.mergeSort(inputArr);
        for(int i:inputArr){
            System.out.print(i);
            System.out.print(" ");
        }
    }

    public void mergeSort(int[] a)
    {
        array = a;
        tempArray = new int[a.length];
        doSort(0, a.length - 1);
    }
    private void doSort(int lowerIndex, int higherIndex)
    {
        if (lowerIndex < higherIndex)
        {
            int middleIndex = (lowerIndex + higherIndex) / 2;
            doSort(lowerIndex, middleIndex);
            doSort(middleIndex + 1, higherIndex);
            mergeParts(lowerIndex, middleIndex, higherIndex);
        }
    }
    private void mergeParts(int lowerIndex, int middleIndex, int higherIndex)
    {
        for (int i = lowerIndex; i <= higherIndex; i++)
            tempArray[i] = array[i];
        int i = lowerIndex;
        int j = middleIndex + 1;
        int k = lowerIndex;
        while (i <= middleIndex && j <= higherIndex)
        {
            if (tempArray[i] <= tempArray[j])
            {
                array[k] = tempArray[i];
                i++;
            }
            else
            {
                array[k] = array[j];
                j++;
            }
            k++;
        }
        while (i <= middleIndex)
        {
            array[k] = tempArray[i];
            i++;
            k++;
        }
    }
}

File: HW3/MergeSortMain/MergeSort.cpp
// MergeSort.cpp : Defines the entry point for the console application.
//

void mergeParts(int lowerIndex, int middleIndex, int higherIndex, int a[])
{
	int *temp = new int[higherIndex - lowerIndex + 1];
	int i = lowerIndex;
	int j = middleIndex + 1;
	int k = 0;

	while (i <= middleIndex && j <= higherIndex)
	{
		if (a[i] <= a[j])
		{
			temp[k] = a[i];
			i++;
		}
		else
		{
			temp[k] = a[j];
			j++;
		}
		k++;
	}
	while (i <= middleIndex)
	{
		temp[k] = a[i];
		i++;
		k++;
	}
	while (j <= higherIndex)
	{
		temp[k] = a[j];
		j++;
		k++;
	}
	for (int i = lowerIndex; i <= higherIndex; i++)
	{
		a[i] = temp[i - lowerIndex];
	}

	delete[] temp;
}

void doSort(int lowerIndex, int higherIndex, int a[])
{
	if (lowerIndex < higherIndex)
	{
		int middleIndex = (lowerIndex + higherIndex) / 2;
		doSort(lowerIndex, middleIndex, a);
		doSort(middleIndex + 1, higherIndex, a);
		mergeParts(lowerIndex, middleIndex, higherIndex, a);
	}
}


void mergeSort(int a[], int n)
{
	doSort(0, n - 1, a);
}


